Add histc op

qili93 opened the pull request:
#### Required（必填, multiple choices, two at most）
- **PR type（PR 类型） is ( A ):**
A. New features（新功能）---------------- D. Performance optimization（性能优化）
B. Bug fixes（问题修复）------------------ E. Breaking changes（向后不兼容的改变）
C. Function optimization（功能优化）------F.  Others（其它）

- **PR changes（改动点）is ( A ):**
A. OPs（operators）---------------------- C. Docs（文档）
B. APIs（接口）--------------------------- D. Others（其它）

- **Use one sentence to describe what this PR does.（简述本次PR的目的和改动）**
Add new operator "paddle.histc"
-----------------------
#### Optional（选填, If None, please delete it）

Fluid Doc PR: https://github.com/PaddlePaddle/FluidDoc/pull/2147


**Add histc op**

Computes the histogram of a tensor. The elements are sorted into equal width bins between min and max. If min and max are both zero, the minimum and maximum values of the data are used.

**Args:**

input (Variable): A Tensor(or LoDTensor) with shape :math:`[N_1, N_2,..., N_k]` . The data type of the input Tensor should be float32, float64, int32, int64.
bins (int): number of histogram bins
min (int): lower end of the range (inclusive)
max (int): upper end of the range (inclusive)

**Returns:**

Variable: Tensor or LoDTensor calculated by histc layer. The data type is int64.

**Code Example 1:**

```python
import paddle
import numpy as np
startup_program = paddle.Program()
train_program = paddle.Program()
with paddle.program_guard(train_program, startup_program):
    inputs = paddle.data(name='input', dtype='int32', shape=[2,3])
    output = paddle.histc(inputs, bins=5, min=1, max=5)
    place = paddle.CPUPlace()
    exe = paddle.Executor(place)
    exe.run(startup_program)
    img = np.array([[2, 4, 2], [2, 5, 4]]).astype(np.int32)
    res = exe.run(train_program,
                  feed={'input': img},
                  fetch_list=[output])
    print(np.array(res[0])) # [0,3,0,2,1]
```

**Code Example 2:**
```python
import paddle
import numpy as np
with paddle.imperative.guard(paddle.CPUPlace()):
    inputs_np = np.array([1, 2, 1]).astype(np.float)
    inputs = paddle.imperative.to_variable(inputs_np)
    result = paddle.histc(inputs, bins=4, min=0, max=3)
    print(result) # [0, 2, 1, 0]
```

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/PaddlePaddle/Paddle?pullRequest=24562) <br/>All committers have signed the CLA.

qili93 commented:
Hi @XiaoguangHu01 and @saxon-zh,

All CI test passed expect 2 approval errors, would you please kindly help to review this pull request? If any comments, please let me know.

Thanks
Qi Li

qili93 commented:
Hi @Xreki,

Would you please kindly help to review the code of this new operator?

Thanks
Qi Li

Xreki commented on paddle/fluid/operators/histc_op.cc:
```diff
@@ -0,0 +1,119 @@
+                          "Input (X) of Histc op should not be null."));
+    PADDLE_ENFORCE_EQ(ctx->HasOutput("Out"), true,
+                      platform::errors::InvalidArgument(
+                          "Output (Output) of Histc op should not be null."));
```
输入输出检查，现在统一用`OP_INOUT_CHECK`

qili93 replied:
已经改成OP_INOUT_CHECK

Xreki commented on paddle/fluid/operators/histc_op.cc:
```diff
@@ -0,0 +1,119 @@
+
+ protected:
+  void Apply(GradOpPtr<T> retv) const override {
+    retv->SetType("histc");
```
这个op的反向也是调用histc本身么

qili93 replied:
histc没有反向OP，已经删除了GradOp

Xreki commented on paddle/fluid/operators/histc_op.cu:
```diff
@@ -0,0 +1,181 @@
+}
+
+template <typename T, typename IndexType>
+__device__ static IndexType getBin(T bVal, T minvalue, T maxvalue,
```
函数命名采用`AxxBxx`这种驼峰式，变量名采用`aaa_bbb`方式。

Xreki commented on paddle/fluid/operators/histc_op.cu:
```diff
@@ -0,0 +1,181 @@
+    T* d_input_min;
+    T* d_input_max;
+    cudaMalloc(reinterpret_cast<void**>(&d_input_min), sizeof(T));
+    cudaMalloc(reinterpret_cast<void**>(&d_input_max), sizeof(T));
```
不要在op里面显式调用`cudaMalloc`，可以用临时Tensor，参考：

https://github.com/PaddlePaddle/Paddle/blob/560c815390f37c818e1f47fe5e8913fa5cd7f54a/paddle/fluid/operators/conv_cudnn_op.cu#L182-L184

Xreki commented on paddle/fluid/operators/histc_op.cu:
```diff
@@ -0,0 +1,181 @@
+                    stream);
+
+    T output_min = (T)minval;
+    T output_max = (T)maxval;
```
数据类型转换用`static_cast`

Xreki commented on paddle/fluid/operators/histc_op.h:
```diff
@@ -0,0 +1,85 @@
+    auto input_numel = input->numel();
+
+    auto output_min = (T)minval;
+    auto output_max = (T)maxval;
```
数据类型转换用`static_cast`

Xreki commented on python/paddle/fluid/tests/unittests/test_histc_op.py:
```diff
@@ -0,0 +1,92 @@
+        self.in_shape = (2, 3)
+        self.bins = 5
+        self.min = 1
+        self.max = 5
```
1. op单测要求每个测试用例输入shape个维度的乘积大于100。
2. TestHistcOpCase1和TestHistcOp配置上没有区别吧？
3. 虽然这个op本身没有grad，但从实现来看，grad也就是histc op本身，那么也需要实现test_grad。若该op确实没有grad，则不需要注册`HistcOpGradMaker`。

Xreki commented on python/paddle/fluid/tests/unittests/test_histc_op.py:
```diff
@@ -0,0 +1,92 @@
+        self.init_test_case()
+        self.inputs = {"X": np.random.random(self.in_shape).astype("float64")}
+        self.init_attrs()
+        Out = np.zeros((self.bins, ), dtype=np.int64)
```
python的计算结果是`全0`？因为`np.random.random`初始化数据的范围都在[0, 1)之间吧，那需要调整下初始化数据的范围，单测不能是这种固定结果的case。

Xreki commented on python/paddle/tensor/linalg.py:
```diff
@@ -751,3 +752,52 @@ def bmm(x, y, name=None):
+          exe.run(fluid.default_startup_program())
+          img = np.array([[2, 4, 2], [2, 5, 4]]).astype(np.int32)
+          res = exe.run(fluid.default_main_program(), feed={'input':img}, fetch_list=[output])
+          print(res) # [0,3,0,2,1]
```
代码示例用动态图的。

Xreki commented on python/paddle/tensor/linalg.py:
```diff
@@ -751,3 +752,52 @@ def bmm(x, y, name=None):
+          img = np.array([[2, 4, 2], [2, 5, 4]]).astype(np.int32)
+          res = exe.run(fluid.default_main_program(), feed={'input':img}, fetch_list=[output])
+          print(res) # [0,3,0,2,1]
+    """
```
用`if in_dygraph_mode()`支持动态图，可参考下别的api实现。

Xreki commented on python/paddle/tensor/linalg.py:
```diff
@@ -751,3 +752,52 @@ def bmm(x, y, name=None):
+    check_type(input, 'X', (Variable), 'histc')
+    dtype = helper.input_dtype()
+    check_dtype(dtype, 'X',
+                ['int32', 'int64', 'float16', 'float32', 'float64'], 'histc')
```
Line 788和Line 790的检查可以使用`check_variable_and_dtype`，另外这个op没有注册`float16`类型。

Xreki commented on python/paddle/tensor/linalg.py:
```diff
@@ -751,3 +752,52 @@ def bmm(x, y, name=None):
+                ['int32', 'int64', 'float16', 'float32', 'float64'], 'histc')
+    check_type(bins, 'bins', (int), 'histc')
+    check_type(min, 'min', (int), 'histc')
+    check_type(max, 'max', (int), 'histc')
```
非Variable类型的参数不需要检查。

Xreki commented on python/paddle/tensor/linalg.py:
```diff
@@ -751,3 +752,52 @@ def bmm(x, y, name=None):
+    check_type(bins, 'bins', (int), 'histc')
+    check_type(min, 'min', (int), 'histc')
+    check_type(max, 'max', (int), 'histc')
+    out = helper.create_variable_for_type_inference(dtype)
```
这个op输入和输入的数据类型不相同，out的数据类型固定为int64，这里应该用`VarDesc.VarType.INT64`。

Xreki commented on paddle/fluid/operators/histc_op.cu:
```diff
@@ -0,0 +1,181 @@
+}
+
+template <typename DeviceContext, typename T>
+class HistcCUDAKernel : public framework::OpKernel<T> {
```
这里可以使用`CUDADeviceContext`特化`HistcKernel`模板类，不需要定义一个新的类。

qili93 commented:
Hi @Xreki and @cyj1986,

Thank you for your review comments! And addressed these comments in the latest new commit, please let me know if any further comments :)

Thanks
Qi Li

qili93 commented:
FluidDoc PR is here: https://github.com/PaddlePaddle/FluidDoc/pull/2147

saxon-zh commented on python/paddle/tensor/linalg.py:
```diff
@@ -751,3 +752,43 @@ def bmm(x, y, name=None):
+          import paddle
+          import paddle.fluid as fluid
+          import numpy as np
+          with fluid.dygraph.guard(fluid.CPUPlace()):
```
使用2.0的API进行示例代码编写吧，paddle.imperative.guard，paddle.imperative.to_variable，paddle.CPUPlace()

qili93 replied:
Addressed in the latest commit :)

Xreki commented on paddle/fluid/operators/histogram_op.cu:
```diff
@@ -0,0 +1,146 @@
+}
+
+template <typename T, typename IndexType>
+__device__ static IndexType GetBin(T bVal, T minvalue, T maxvalue,
```
变量名`bVal`还是建议按照[google c++代码风格](https://zh-google-styleguide.readthedocs.io/en/latest/google-cpp-styleguide/naming/)来，写成`b_val`。下同。

Xreki commented on python/paddle/fluid/tests/unittests/test_histogram_op.py:
```diff
@@ -0,0 +1,87 @@
+
+    def test_check_output(self):
+        self.check_output()
+
```
没有报错相关的单测。

Xreki commented on python/paddle/tensor/linalg.py:
```diff
@@ -751,3 +752,63 @@ def bmm(x, y, name=None):
+    Returns:
+        Variable: Tensor or LoDTensor calculated by histogram layer. The data type is int64.
+
+    Code Example 1:
```
这个示例可以删除了。另外，这么写，英文文档可以正常预览么？

Xreki commented on python/paddle/tensor/linalg.py:
```diff
@@ -751,3 +752,63 @@ def bmm(x, y, name=None):
+                print(np.array(res[0])) # [0,3,0,2,1]
+
+    Code Example 2:
+        .. code-block:: python
```
`code-block`下面要有一个空行。

Xreki commented on paddle/fluid/operators/histogram_op.cu:
```diff
@@ -0,0 +1,146 @@
+      const IndexType bin =
+          GetBin<T, IndexType>(inputVal, minvalue, maxvalue, nbins);
+      const IndexType outputIdx = bin < nbins - 1 ? bin : nbins - 1;
+      paddle::platform::CudaAtomicAdd(&output[outputIdx], 1);
```
用原子操作更新，估计会比较慢。这个kernel应该可以使用shared memory优化，不过可以下个PR再考虑。

Xreki commented on python/paddle/fluid/tests/unittests/test_histogram_op.py:
```diff
@@ -0,0 +1,87 @@
+    def setUp(self):
+        self.op_type = "histogram"
+        self.init_test_case()
+        np_input = np.random.randint(
```
单测中，输入数据类型要求是fp64。

Xreki commented on python/paddle/tensor/linalg.py:
```diff
@@ -751,3 +752,63 @@ def bmm(x, y, name=None):
+def histogram(input, bins=100, min=0, max=0):
+    """
+    Computes the histogram of a tensor. The elements are sorted into equal width bins between min and max. 
+    If min and max are both zero, the minimum and maximum values of the data are used.
```
浮点数按四舍五入范围归类么，能否说明下。

Xreki commented on python/paddle/tensor/linalg.py:
```diff
@@ -751,3 +752,63 @@ def bmm(x, y, name=None):
+            with paddle.imperative.guard(paddle.CPUPlace()):
+                inputs_np = np.array([1, 2, 1]).astype(np.float)
+                inputs = paddle.imperative.to_variable(inputs_np)
+                result = paddle.histogram(inputs, bins=4, min=0, max=3)
```
能否举个浮点数的例子。

qili93 commented:
在下个PR中address以下内容:

1. paddle/fluid/operators/histogram_op.cu
    1.1 变量名修改，例如bVal => b_val 
    1.2 kernel使用shared memory优化CudaAtomicAdd
2. python/paddle/tensor/linalg.py 修改:
    2.1 code-block下面要有一个空行
    2.2 确认英文文档可以正常预览
3. python/paddle/fluid/tests/unittests/test_histogram_op.py
    3.1 添加报错相关的单测
    3.2 单测中，输入数据类型要求是fp64
    3.3 增加浮点数的例子，说明浮点数计算规则
